src/NuGetForUnity/Editor/NugetWindow.cs

```diff
@@ -724,23 +731,27 @@ private void DrawUpdatesHeader()
                         UpdateUpdatePackages();
                     }
 
-                    if (GUILayout.Button("Install All Updates", GUILayout.Width(150)))
+                    if (updatePackages.Count > 0)
                     {
-                        NugetHelper.UpdateAll(updatePackages, NugetHelper.InstalledPackages);
-                        NugetHelper.UpdateInstalledPackages();
-                        UpdateUpdatePackages();
-                    }
+                        if (GUILayout.Button("Update All", GUILayout.Width(150)))
+                        {
+                            NugetHelper.UpdateAll(updatePackages, NugetHelper.InstalledPackages);
+                            NugetHelper.UpdateInstalledPackages();
+                            UpdateUpdatePackages();
+                        }
 
-                    if (GUILayout.Button("Refresh", GUILayout.Width(60)))
-                    {
-                        Refresh(true);
+                        if (updatePackages.Where(p => p.IsSelected).Count() > 0)
+                        {
+                            if (GUILayout.Button("Update Selected", GUILayout.Width(200)))
+                            {
+                                NugetHelper.UpdateAll(updatePackages.Where(p => p.IsSelected), NugetHelper.InstalledPackages);
```

Comment:
I still think a private field `private readonly HashSet<NugetPackage> selectedPackages = new HashSet<NugetPackage>();`. Then wenn you need the selected packages you need to do: `updatePackages.Where(selectedPackages.Contains)`. The reason for this is just to separate UI state from logic. E.g. when refreshing the data (updatePackages) you don't automatically lose the selection state (what package the user has selected).

Reply:
Ah, right. Just to clearify: it should be better if there is no IsSelected field in NugetPackage itself since that is view only data and should thus only exist in the view layer and one way to save what are the selected packages in view layer is to use `HashSet<NugetPackage>` there. Is that right?

Reply:
Exactly